jbmc/src/java_bytecode/lambda_synthesis.cpp

```diff
@@ -610,6 +610,82 @@ static symbol_exprt instantiate_new_object(
   return new_instance_var;
 }
 
+/// If \p expr needs (un)boxing to satisfy \p required_type, add the required
+/// symbols to \p symbol_table and code to \p code_block, then return an
+/// expression giving the adjusted expression. Otherwise return \p expr
+/// unchanged. \p role is a suggested name prefix for any temporary variable
+/// needed; \p function_id is the id of the function any created code it
+/// added to.
+exprt adjust_type_if_necessary(
+  exprt expr,
+  const typet &required_type,
+  code_blockt &code_block,
+  symbol_table_baset &symbol_table,
+  const irep_idt &function_id,
+  const std::string &role)
+{
+  const typet &original_type = expr.type();
+  const bool original_is_pointer = can_cast_type<pointer_typet>(original_type);
+  const bool required_is_pointer = can_cast_type<pointer_typet>(required_type);
+
+  if(original_is_pointer == required_is_pointer)
+  {
+    if(original_is_pointer && original_type != required_type)
+      return typecast_exprt{expr, required_type};
+    else
+      return expr;
+  }
+
```

Comment:
What's this case for?

Reply:
I think the answer is byte to int and Derived -> Base?

Reply:
Gains a comment in a later commit (it's for pointer or integer type clashes without boxing, where plain casting always does the trick (forcing pointers and widening integers, so long as only valid Java casts are fed in in the first place)